### DISCUSSION TITLE
From raw data to database and the UniqueObject recipe

### DISCUSSION THREAD

**Discussion by MateoSaezMata** [sqlalchemy/sqlalchemy#10168] [Usage Questions]
## Some context
I'm working as data engineer in some projects related with dairy production. At this industry the data that we are collecting is always delivered in an unstructured way, such csvs or xlsx that are autogenerated by some software whose goal is handle the production itself (in closed source manner).
## Our goal
Leaving aside the fact that we are storing this csvs files into our NoSQL storage, we are trying to structure this data into an SQL database through python. That's where **SQLAlchemy's ORM** comes into place.
## Raw data
The data provided shows the dairy production from some animals (cows) across different locations (farms).

So we recive a csv file every day whose fields are:
**[ date | farm | farm_location | cow | cow_breed | production ]**

To be clear as possible let's set a fully example of two days of raw data:

**27Jul'23 data**
`20230727_prod.csv`
| date       	| farm   	| farm_location 	| cow   	| cow_breed   	| production 	|
|------------	|--------	|---------------	|-------	|-------------	|------------	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_1 	| Brown Swiss 	| 40         	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_2 	| Jersey      	| 38         	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_3 	| Jersey      	| 35         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_4 	| Jersey      	| 42         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_5 	| Brown Swiss 	| 37         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_6 	| Brown Swiss 	| 33         	|

**28Jul'23 data**
`20230728_prod.csv`
| date       	| farm   	| farm_location 	| cow   	| cow_breed   	| production 	|
|------------	|--------	|---------------	|-------	|-------------	|------------	|
| 28/07/2023 	| farm_1 	| Montana       	| cow_1 	| Brown Swiss 	| 35         	|
| 28/07/2023 	| farm_2 	| Nevada          | cow_4 	| Jersey      	| 41         	|
| 28/07/2023 	| farm_2 	| Nevada       	| cow_5 	| Brown Swiss  | 36         	|
| 28/07/2023 	| farm_2 	| Nevada        	| cow_6 	| Brown Swiss   | 36         	|
| 28/07/2023 	| farm_1 	| Montana        | cow_7 	| Jersey    	| 38         	|
| 28/07/2023 	| farm_2 	| Nevada        	| cow_8 	| Brown Swiss 	| 43         	|

#### Notes on the data
- The animals are present at the collected data if they had produced some product (milk) at the specific day.
- New animals can become part of the production.
## The SQLAlchemy models
Given the previous data we decided to structure it into three tables. The schema is as follows:
<p align="center">

<img src="https://github.com/sqlalchemy/sqlalchemy/assets/111874229/8e109ed2-f4e0-42ad-b907-3e971a4c86ed" alt="drawing" width="160"/>
</p>

And the SQLAlchemy models code:
```python
from __future__ import annotations
from typing import List
import datetime
from sqlalchemy.orm import (
    Mapped,
    mapped_column,
    relationship,
    DeclarativeBase
)
from sqlalchemy import ForeignKey

class Base(DeclarativeBase):
    pass

class Farm(Base):
    __tablename__ = 'farm'

    farm_id: Mapped[str] = mapped_column(primary_key=True)
    location: Mapped[str]

    # ONE TO MANY
    cows: Mapped[List[Cow]] = relationship(back_populates='farm')

class Cow(Base):
    __tablename__ = 'cow'

    cow_id: Mapped[str] = mapped_column(primary_key=True)
    farm_id: Mapped[str] = mapped_column(ForeignKey('farm.farm_id'))
    cow_breed: Mapped[str]

    # ONE TO MANY
    productions: Mapped[List[Production]] = relationship(back_populates='cow')
    # MANY TO ONE
    farm: Mapped[Farm] = relationship(back_populates='cows')

class Production(Base):
    __tablename__ = 'production'

    date: Mapped[datetime.date] = mapped_column(primary_key=True)
    cow_id: Mapped[str] = mapped_column(ForeignKey('cow.cow_id'), primary_key=True)
    quantity: Mapped[int] # This is equivalent to production field on raw data

    cows: Mapped[Cow] = relationship(back_populates='productions')
```
## Database inserting approach
First, we define all the previous resources:
```python
from sqlalchemy import (
    create_engine,
    URL
)
from sqlalchemy.orm import (
    sessionmaker
)
from models_example import (
    Base,
    Farm,
    Cow,
    Production
)
import pandas as pd

DRIVER_NAME = 'postgresql'
USERNAME = '<>'
PWD = '<>'
HOST = 'localhost'
PORT = '5432'
DATABASE = 'example_db'
ECHO = True

url_object = URL.create(
    drivername=DRIVER_NAME,
    username=USERNAME,
    password=PWD,
    host=HOST,
    port=PORT,
    database=DATABASE
)

engine = create_engine(url_object, echo=ECHO)

Session = sessionmaker(bind=engine)

sess = Session()
```
And then, we iterate through the DataFrame to read the data.
For me, makes sense instantiate the objects as follows:
(I don't know what option is better. I'm assuming that the first one is possible more clear)
#### OPTION1:
```python
raw_data = pd.read_csv('<file_path>') # '20230727_prod.csv' , '20230728_prod.csv' , etc ...

for index, row in raw_data.iterrows():
    f = Farm(
        farm_id=row['farm'],
        location=row['location']
            )
    c = Cow(
        cow_id=row['cow'],
        cow_breed=row['cow_breed'],
        farm=f
            )
    p = Production(
        date=row['date'],
        quantity=row['production'],
        cow=c
    )
    sess.add(p)
```
#### OPTION2:
```python
raw_data = pd.read_csv('<file_path>') # '20230727_prod.csv' , '20230728_prod.csv' , etc ...

for index, row in raw_data.iterrows():
    t = Production(
        date=row['date'],
        quantity=row['production'],
        cow=Cow(
            cow_id=row['cow'],
            cow_breed=row['cow_breed'],
            farm=Farm(
                farm_id=row['farm'],
                location=row['location']
            )
        )
    )
    sess.add(t)
```

#### NOTES:
(I'm trying to put my thoughts explicitly to be clear as possible)
- The `df.iterrows()` method isn't the best "performance way" to read the data but this is another issue (related with another library). Anyway, if someone knows a better way to instantiate objects from a df it will be good to know.
- When we `sess.add(p)` at option 1 we are adding the `Production` object to the session. Since we defined the relationships along the models, the cascade adds all the related objects (`Farm` and `Cow`) to the session. Same behavior is expected at option 2. From here the session changes all objects' state from `transient` to `pending`. The transaction's scope hasn't started yet.
## Business logic
From a business logic point of view, we need our tables to grow from the time a cow is producing within a location. So we need the Farm-Cow-Production combo works as a whole.
Using other words: when we go into the for loop iteration, each iteration must works as a whole, it must be atomic.
**DOUBT-1**
Is this mean that at every iteration we should perform a `sess.commit()`?
I think this is the correct approach but I've some concerns because the performance of writing (in a persistence way) at every iteration.
## The duplicates issue
Probably someone has been able to notice that due to the above script, we have been instantiating duplicated objects. Lets go through the code that applies at the 27Jul'23 production data (`20230727_prod.csv`)
| date       	| farm   	| farm_location 	| cow   	| cow_breed   	| production 	|
|------------	|--------	|---------------	|-------	|-------------	|------------	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_1 	| Brown Swiss 	| 40         	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_2 	| Jersey      	| 38         	|
| 27/07/2023 	| farm_1 	| Montana       	| cow_3 	| Jersey      	| 35         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_4 	| Jersey      	| 42         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_5 	| Brown Swiss 	| 37         	|
| 27/07/2023 	| farm_2 	| Nevada        	| cow_6 	| Brown Swiss 	| 33         	|

At first iteration:
```python
f1 = Farm(
    farm_id='farm_1',
    location='Montana'
    )
c1 = Cow(
    cow_id='cow_1',
    cow_breed='Brown Swiss',
    farm=f1
    )
p1 = Production(
    date='27/07/2023', # This shouldn't be a string (it isn't) but for the example is ok
    quantity=40,
    cow=c1
    )
sess.add(p1)
```
At the second:
```python
f2 = Farm(
    farm_id='farm_1',
    location='Montana'
    )
c2 = Cow(
    cow_id='cow_2',
    cow_breed='Jersey',
    farm=f2
    )
p2 = Production(
    date='27/07/2023', # This shouldn't be a string (it isn't) but for the example is ok
    quantity=38,
    cow=c2
    )
sess.add(p2)
```
We can easily check that, from a business logic point of view, `f1` represents the same row that `f2`; but from an SA/python point of view, this two objects aren't the same (they are different objects at different memory location).

Needless to say, our goal will be to insert this record once at the database.

After reading a lot on the docs and watch every single explanation that @zzzeek has made on the SA's philosophy I could understand that this behavior (handling same object) is managed during a transaction by the `identity map`. 

Since we have only added objects to the session and the transaction has not yet started, the library has not been able to effectively identify which objects, of those that were pending, already existed to be inserted.

The above thoughts are related with duplicated objects instantiated during the same execution runtime. But the same logic applies if the object we are instantiating already exists in permanent way at the database.
## The UniqueObject recipe
Talking about the possible solutions we need to talk about the [UniqueObject](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) recipe. This implementation is intended to handle duplicates at declaration time. Let's get the code from the SA wiki:
```python
def _unique(session, cls, hashfunc, queryfunc, constructor, arg, kw):
    cache = getattr(session, '_unique_cache', None)
    if cache is None:
        session._unique_cache = cache = {}

    key = (cls, hashfunc(*arg, **kw))
    if key in cache:
        return cache[key]
    else:
        with session.no_autoflush:
            q = session.query(cls)
            q = queryfunc(q, *arg, **kw)
            obj = q.first()
            if not obj:
                obj = constructor(*arg, **kw)
                session.add(obj)
        cache[key] = obj
        return obj
```
As the description at the link says:

> We'll define a function _unique() that will provide the "guts" to the unique recipe. This function is given a Session to work with, and associates a dictionary with the Session() which keeps track of current "unique" keys.

I try to explain (as I understand it) the logic flow of the function:
1. Get the `_unique_cache` dict from the session if exists, otherwise assigns the dict as property of the `session`.
2. Create the `key` as unique identifier of the object.
3. Check if `key` already exists at `_unique_cache` and returns the existing object if possible.
4. If not, look for the object querying the database and if not instantiate it through the regular constructor.
5. Regardless of whether we have it in the database or we have to construct it, we add it to the cache.

So, through this logic we solve above mentioned issues: get unique objects between runtime construction and database persistence.
### Automating `hashfunc` and `queryfunc`
The `hashfunc` and `queryfunc` are implemented at the UniqueObject example, in a similar manner but with different goals:
- `hashfunc` purpose is to create a unique key which output is **some unique `field: value` pairs**
- `queryfunc` purpose is to query the DB looking for the existence of a row which satisfies the **same unique `field: value` pairs**

The way I would like to implement these two functions would be to automatically find the table fields that have a unique constraint (**usually the primary key fields**) and let these two functions generate the `key` and the `query ` to the database using the field and value pairs that match those found in the unique constraints (**primary keys**).

We enhance the `Base` class to include the `as_unique` classmethod:

```python
class Base(DeclarativeBase):

    @classmethod
    def as_unique(cls, session: Session, *arg, **kw):
        cache = getattr(session, '_unique_cache', None)
        if cache is None:
            session._unique_cache = cache = {}
        
        pks = [col.name for col in cls.__table__.primary_key] # Get the PKs names
        pk_kw = {}
        for k, v in kw.items(): 
            if k in pks:
                pk_kw[k] = v # Get the PK value from **kw
                
        key = cls.__mapper__.identity_key_from_primary_key(pk_kw.values()) # Generate the unique key
        if key in cache:
            return cache[key]
        else:
            with session.no_autoflush:
                obj = session.query(cls).filter_by(**pk_kw).first() # Query the DB using filter_by + pk_kw key-value pairs
                if not obj:
                    obj = cls(*arg, **kw)
            cache[key] = obj
            return obj
```
So far so good, I've tested this implementation and works as expected.

## The relationship property issue
Seems that all is working as expected so far but we already have a problem:
We are trying to find out the field-value pairs automatically, looking for them on the property `cls.__table__.primary_key` (that is ok) but when we create/instantiate our objects initially we are using relationship properties feature (a nice one by the way).
Looking at the code again:

The `Production` class/table has a composite primary key: `[date, cow_id]`
```python
class Production(Base):
    __tablename__ = 'production'

    date: Mapped[datetime.date] = mapped_column(primary_key=True)
    cow_id: Mapped[str] = mapped_column(ForeignKey('cow.cow_id'), primary_key=True)
    quantity: Mapped[int] # This is equivalent to production field on raw data

    cow: Mapped[Cow] = relationship(back_populates='productions')
```
And when we instantiate `Production`objects:
```python
f1 = Farm(
    farm_id='farm_1',
    location='Montana'
    )
c1 = Cow(
    cow_id='cow_1',
    cow_breed='Brown Swiss',
    farm=f1
    )
p1 = Production(
    date='27/07/2023', # This shouldn't be a string (it isn't) but for the example is ok
    quantity=40,
    cow=c1
    )
sess.add(p1)
```
We are defining them using `date`, `quantity` and **`cow` relationship property**.
**That is the point**: since we are not setting the `cow_id` property/field of the object, this value is set to None until some flush happens during a transaction (I think).

Our above `as_unique` method tries to find the field-value pairs corresponding to a primary key fields, but since we are not defined `cow_id` the functionality of the `as_unique` method doesn't work as expected anymore.

I've tried to implement some functionality that solves that issue, looking for through the relationship property of the class but seems a lot weird to me and I don't know the library enough to be confident on this implementation.
## Conclusions
I recently start to learn SQLAlchemy and I like the way that solves the operations against a database. Because of this probably I am misunderstanding something.
Probably is a better way to solve this task I don't know. At this moment all the above has pretty sense to me, except for the UniqueObject recipe when relationships come into place and I am trying to automate hash and query funcs.
Thanks in advance to someone who could help me to understand my mistakes :)

**Comment by zzzeek** [comment]
hi -

I'm not the one to make a definitive response to this since this looks like more of a lot of notes about an implementation in progress.

The one suggestion I'd make immediately though is to ensure the database has unique constraints for these various fields which you want to be unique.   it should not be possible to INSERT duplicate rows in the database; instead, it should throw an error.

once you have dupes throwing unique constraint violations, you have the option to use optimistic systems to INSERT these rows, catch a unique violation, then skip that row (using a savepoint).   you then have no need to worry about "atomic".

**Reply by MateoSaezMata** [reply]
Hi! Thanks for the quick reply.
I tried to be clear as possible so that anyone who reads the subject would fully understand the problem.
The short answer would be: 
**Is possible to use the [UniqueObject](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) recipe with an object that is constructed using relationship properties, and this relationship is related with a unique constrain field?**
The optimistic approach probably includes an hybrid solution where on one side we can use `as_unique` method on rows that have a high probability to already be on the database and on the other side we can use the standard way of construction on rows that probably won't be on the database and catch the unique error on those cases.
I let you know how I solved the issue.

Thanks

**Comment by CaselIT** [comment]
Hi,

Depending on the volume of data to insert the following approach may be a lot faster, since you already have all the primary keys:
- read the data as a pandas like you are doing
- select just farm and location from the original datafrane
  - drop duplicates on farm id
  - convert the datafrane to a list of dict using `to_dict(orient='records')`
  - use insert on conflict do nothing to insert all the missing farms https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_nothing
- starting from the original datafrane do the same thing for the cows (select cowid, breed, farm; drop duplicates on the id; insert the dicts
- select just the daily production columns on the original datafrane and insert it using a plain insert of the list of dicts.

While the above procedure does not use orm features, it's likely singificantly faster. The orm classes can be used to query the data.

**Reply by MateoSaezMata** [reply]
Hi! Thanks for answering!
We are already applying techniques like you are describing.
(Short description) 
- First we are filtering the fields on the Dataframe related with the target table.
- Then query the database table and store it into another DataFrame.
-  Then dropping duplicates based on incoming (csv) and queried (db) data.
- Then insert new data.

The main difference is that we are querying the database instead using the ON CONFLICT DO NOTHING.
We are moving into SQLAlchemy ORM chasing more explicit (declarative) approaches on business logic (since all the drop duplicates issue isn't business logic I think), because some problem different on Unique Constraint Violation could arise.
The goal is prioritize data quality versus performance (at least up to a point; we are handling at this moment 1GB datasets at most).
Never used an INSERT with the clause ON CONFLICT DO NOTHING, I'll take a look on it.

Thanks for your reply

**Reply by CaselIT** [reply]
Sqlalchemy allows you do mix and match use of the orm with use of the core parts (that cleanly translates to sql), so you can choose what's the most suited approach
